src/server/game/Spells/Auras/SpellAuras.cpp

```diff
@@ -1672,18 +1672,33 @@ bool Aura::CanStackWith(Aura const* existingAura) const
     return true;
 }
 
-bool Aura::IsProcOnCooldown(std::chrono::steady_clock::time_point now) const
+bool Aura::IsProcOnCooldown(std::chrono::steady_clock::time_point now, ObjectGuid procTarget) const
 {
+    if (GetSpellInfo()->HasAttribute(SPELL_ATTR10_PROC_COOLDOWN_PER_TARGET))
+    {
+        auto itr = m_targetProcCooldown.find(procTarget);
+        if (itr != m_targetProcCooldown.end())
+            return itr->second > now;
+
+        // Not found in the target map, should not proc.
+        return false;
+    }
+
     return m_procCooldown > now;
 }
 
-void Aura::AddProcCooldown(std::chrono::steady_clock::time_point cooldownEnd)
+void Aura::AddProcCooldown(std::chrono::steady_clock::time_point cooldownEnd, ObjectGuid procTarget)
 {
-    m_procCooldown = cooldownEnd;
+    if (GetSpellInfo()->HasAttribute(SPELL_ATTR10_PROC_COOLDOWN_PER_TARGET))
```

Comment:
you gotta check for procTarget.IsEmpty() as well, otherwise you might store cooldown data for  m_targetProcCooldown[ObjectGuid::Empty] if no proc target is given but the spell has  SPELL_ATTR10_PROC_COOLDOWN_PER_TARGET

Reply:
I don't like this solution anyways, not sure how blizzard does it ...